bootstrap/005_table_migrations.sql

```diff
@@ -25,35 +25,62 @@ BEGIN
+      SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_SCHEMA = :table_schema AND TABLE_NAME = :table_name
       )
+
       );
-  IF (missing_view_columns) THEN
-      SYSTEM$ADD_EVENT('table out of sync - columns exist in table distinct from those in view', {'count_missing': (:missing_view_columns)});
-      RAISE OUT_OF_SYNC;
-  END IF;
 
-  let columns string := (
-      SELECT LISTAGG('"' || COLUMN_NAME || '" ' || DATA_TYPE, ', ') WITHIN GROUP (ORDER BY ORDINAL_POSITION)
+  let columns_to_drop string := (
+      SELECT LISTAGG('"' || COLUMN_NAME || '" ' , ', ')
       FROM (
-      SELECT COLUMN_NAME, DATA_TYPE, ORDINAL_POSITION FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_SCHEMA = :view_schema AND TABLE_NAME = :view_name
+      SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_SCHEMA = :table_schema AND TABLE_NAME = :table_name
       MINUS
-      SELECT COLUMN_NAME, DATA_TYPE, ORDINAL_POSITION FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_SCHEMA = :table_schema AND TABLE_NAME = :table_name
+      SELECT COLUMN_NAME, DATA_TYPE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_SCHEMA = :view_schema AND TABLE_NAME = :view_name
       )
 
       );
-  let alter_statement string := 'ALTER TABLE "' || :table_schema || '"."' || :table_name || '" ADD ' || columns;
-  if (columns <> '') then
-      execute immediate alter_statement;
+
+  let alter_statement string := 'ALTER TABLE "' || :table_schema || '"."' || :table_name || '"';
+
+  let alter_table_add_column string := '';
+  let alter_table_drop_column string := '';
+
+  if (columns_to_add <> '') then
+    alter_table_add_column := ' ADD ' || columns_to_add;
+    execute immediate alter_statement || alter_table_add_column;
+  end if;
+
+  if (columns_to_drop <> '') then
+    alter_table_drop_column := ' DROP ' || columns_to_drop;
+    execute immediate alter_statement || alter_table_drop_column;
+  end if;
+
+  if (columns_to_add <> '' OR columns_to_drop <> '') then
       SYSTEM$LOG_INFO('Migration executed for ' || :view_schema || '.' || :view_name || ' and ' || :table_schema || '.' || :table_name);
-      SYSTEM$ADD_EVENT('table altered', {'alter_statement': alter_statement });
-      RETURN columns;
+      SYSTEM$ADD_EVENT('table altered', {'alter_statement': alter_statement || alter_table_add_column || alter_table_drop_column });
+      RETURN alter_table_add_column || alter_table_drop_column;
   else
     SYSTEM$LOG_INFO('No migration need for ' || :view_schema || '.' || :view_name || ' and ' || :table_schema || '.' || :table_name);
     RETURN null;
   end if;
 END;
+
+
+CREATE OR REPLACE PROCEDURE internal.migrate_view()
+    RETURNS STRING
+    LANGUAGE SQL
+    COMMENT = 'Re-create view used in the App. This is required when Snowflake adds a new column to query_history, or removes an existing column from query_history'
+    AS
+BEGIN
+    call INTERNAL.create_view_QUERY_HISTORY_COMPLETE_AND_DAILY();
```

Comment:
where are these procs defined?

Reply:
ah, i see it now. Does this work ok in a fresh install of opscenter? We are calling stored procs from `005` which arent defined till `016`.

Also, why don't we have to do the same for the warehouses tables?

Reply:
I did both `deploy/deploy.py` and `deploy/devdeploy.py`, and  both cases run fine.

For warehouse tables, we use `select * from INTERNAL_REPORTING.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY ` when create table.    The view  `INTERNAL_REPORTING.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY`  uses explicit column list. If the underlying `ACCOUNT_USAGE.WAREHOUSE_EVENTS_HISTORY` is changed, we do not have to re-create the tables.

Reply:
but we have the same problem as query history? if they remove a column from the table and we haven't released opscenter w/ the changes then the app will be broken?

It seems like this migration pattern is generic for all table types we rely on

Reply:
I saw we have the following select clause over `warehouse_events_history`.  If the column explicitly referred in the expressions are dropped by Snowflake, we have to change the view DDL in OpsCenter, and redeploy with the modified view DDL.  If Snowflake adds a new column, or removes a column that is not referred in the expression here, then OpsCenter will work as it is without making any migration/change.

In other words, if the columns explicitly used here are dropped by Snowflake, I feel we have to update OpsCenter app, and redeploy it. 

```
    SELECT
        timestamp as filterts,
        timestamp,
        warehouse_id,
        warehouse_name,
        cluster_number,
        case
            when event_name in ('RESUME_WAREHOUSE','SUSPEND_WAREHOUSE') then 'WH'
            when event_name in ('RESUME_CLUSTER','SUSPEND_CLUSTER') then 'CL'
        END as SESSION_TYPE,
        case
            when event_name = 'RESUME_WAREHOUSE' and event_state = 'STARTED' then 1
            when event_name = 'RESUME_CLUSTER' and event_state = 'COMPLETED' then 1
            when event_name = 'SUSPEND_CLUSTER' and event_state = 'COMPLETED' then 2
            when event_name = 'SUSPEND_WAREHOUSE' and event_state = 'COMPLETED' then 2
            else null
        end as SESSION_BEHAVIOR,
        case
            when event_name = 'RESUME_WAREHOUSE' and event_state = 'STARTED' then 1
            when event_name = 'RESUME_CLUSTER' and event_state = 'COMPLETED' then 2
            when event_name = 'SUSPEND_CLUSTER' and event_state = 'COMPLETED' then 3
            when event_name = 'SUSPEND_WAREHOUSE' and event_state = 'COMPLETED' then 4
            else null
        end as EVENT
    FROM ACCOUNT_USAGE.WAREHOUSE_EVENTS_HISTORY
```

Reply:
> did you drop the db before running teh above?

I did not drop the db before the above `deploy/deploy.py`, and `deploy/devdeploy.py`.   I just re-deploy after dropping the App and dbs, and they run fine.

```
drop application jinfeng_opscenter_app;
drop database jinfeng_opscenter_pkg;

drop database jinfeng_opscenter;

poetry run python deploy/deploy.py -p jinfeng_local -v V1

poetry run python deploy/devdeploy.py -p local_dev
```